[Core] - Move core-tracing to @opentelemetry/api 0.20.0

maorleger opened the pull request:
## What

- Bump @opentelemetry/api to 0.20.0 in @azure/core-tracing
- Move all packages that are on core-http to the next core-tracing version 
- Remove version collision check from the tracer cache

## Why

This is part of our effort to move everyone to OTel 0.20.0 - but we have to stage it due to a transitive dependency and a breaking change in OTel. This PR updates core-tracing to use the latest OTel, fixes any breaking changes, and moves packages that we can move to the latest version of core-tracing.

Once core-rest-pipeline 1.1.0 is GA'd we'll be able to move the rest of the packages over to the latest core-tracing as well.

Removing the version collision came out of an offline discussion after these changes were reviewed - since it's a common source of pain and hasn't added much benefit (tracer conflicts incompatibility was never a problem) we decided to remove that logic
and always grab a unique symbol per OT compatibility. 

## Callouts

The packages that are already on core-v2 will _not_ be upgraded at this time - we'll coordinate that with the GA of core-rest-pipeline 1.1.0

maorleger commented on common/config/rush/common-versions.json:
```diff
@@ -74,6 +74,11 @@
+
+
+
+    // TODO: remove 12 once it's the common version, and add an issue to remove this entry once everyone is upgraded
```
I won't merge with this TODO in place

maorleger commented on sdk/appconfiguration/app-configuration/package.json:
```diff
@@ -89,13 +89,13 @@
     "@azure/dev-tool": "^1.0.0",
     "@azure/eslint-plugin-azure-sdk": "^3.0.0",
-    "@azure/identity": "2.0.0-beta.3",
+    "@azure/identity": "2.0.0-beta.4",
```
This identity update is necessary because it also has a transitive dependency on core-http. Talking to @sadasant it is a safe update to make, and required to get the new context method call

maorleger commented on sdk/core/core-tracing/review/core-tracing.api.md:
```diff
@@ -85,25 +85,22 @@ export type HrTime = [number, number];
-    traceId: string;
-    spanId: string;
-};
+export type LinkContext = SpanContext;
```
This might not be needed at all, I'll take a look. [1.0.0-rc3](https://unpkg.com/browse/@opentelemetry/api@1.0.0-rc.3/build/src/trace/link.d.ts) you can see that `LinkContext` was a subset of `SpanContext`... in [0.20.0](https://unpkg.com/browse/@opentelemetry/api@0.20.0/build/src/trace/link.d.ts) it was changed to just be the `SpanContext`

maorleger commented on sdk/eventhub/event-hubs/test/internal/sender.spec.ts:
```diff
@@ -725,7 +725,7 @@ describe("EventHub Sender", function(): void {
         ]
       };
 
```
I can't wait to move these test methods to `@azure/test-utils` after this upgrade

maorleger commented on sdk/keyvault/keyvault-keys/test/public/recoverBackupRestore.spec.ts:
```diff
@@ -118,7 +118,10 @@ describe("Keys client - restore keys and recover backups", () => {
-      const restorePoller = await testClient.beginRestoreKeyBackup(backup as Uint8Array);
+      const restorePoller = await testClient.beginRestoreKeyBackup(
+        backup as Uint8Array,
+        testPollerProperties
```
Unrelated to tracing, but brings our recorded test timing down a few orders of magnitude by setting pollInterval to 0 in playback mode 👍

richardpark-msft commented on sdk/core/core-tracing/src/interfaces.ts:
```diff
@@ -260,7 +254,7 @@ export interface Span {
    * @returns the SpanContext object associated with this Span.
    */
-  context(): SpanContext;
+  spanContext(): SpanContext;
```
As per our offline discussion - consider if we even need to expose this at all if we abstract a bit more of our testing code.

maorleger replied:
I do want to do this, but we need to abstract away our testing code first so I created https://github.com/Azure/azure-sdk-for-js/issues/15730 to track this.

richardpark-msft commented:
Looks good (and thankfully the source level changes are minimal, unlike last time).

One other thing is to update for the [core-tracing/cache.ts](https://github.com/Azure/azure-sdk-for-js/blob/2280fd39ca4f4c3f9eb8f50737c34b1b4e2e1653/sdk/core/core-tracing/src/utils/cache.ts#L11).

CC: @xirzec

maorleger commented:
@xirzec @richardpark-msft thanks Richard for the reminder! 

I didn't love that I had to be reminded to bump this version though - so I did some silly things to try to make it better in this commit: https://github.com/Azure/azure-sdk-for-js/pull/15672/commits/9b84b46cc9159641a325c237fe41ba0fef9daca6

Basically, assuming I didn't break _everything_ what I did was:

- match the GLOBAL_TRACER_VERSION to @opentelemetry/api version (made it a string, assuming it can be)
- Inferred the GLOBAL_TRACER_SYMBOL from GLOBAL_TRACER_VERSION to avoid repeating ourselves there
- added a test to ensure that nobody forgets to do this in the future.

But I did some things without thinking or understanding them like adding `json()` to the rollup config or changing the type of cache symbol - so... hopefully you folks will be kind enough to let me know if we can't do this and I'll revert.

deyaaeldeen reviewed (commented):
I reviewed the PR and it looks good to me, thanks for upgrading us!

I will defer to @richardpark-msft for approval.

xirzec commented on sdk/core/core-tracing/src/utils/cache.ts:
```diff
@@ -8,13 +8,14 @@ import { getGlobalObject } from "./global";
 // preview5 shipped with @azure/core-tracing.tracerCache
 // and didn't have smart detection for collisions
-const GLOBAL_TRACER_SYMBOL = Symbol.for("@azure/core-tracing.tracerCache3");
+const GLOBAL_TRACER_SYMBOL = Symbol.for(`@azure/core-tracing.tracerCache${GLOBAL_TRACER_VERSION}`);
```
If you're putting the version into the symbol name, you might as well delete the `else` clause on line 32, because we're never going to warn about incompatible global tracers. You can also remove `GLOBAL_TRACER_VERSION` entirely and delete `version` from the cache interface, since it's superfluous.

The point of the version was to track if the tracer we were storing was from a different and incompatible version of `core-tracing`.

xirzec commented on sdk/core/core-tracing/test/cache.spec.ts:
```diff
@@ -0,0 +1,21 @@
+    fs.readFileSync(path.join(__dirname, "../package.json"), { encoding: "utf8" })
+  );
+
+  it("ensure version matches @opentelemetry/api version", () => {
```
per above comment, we can delete this whole constant and then we don't need a test for it

xirzec reviewed (changes requested):
I'm glad the actual changes are pretty minor!

This also reminded me to file #15732

xirzec commented on sdk/core/core-tracing/src/utils/cache.ts:
```diff
@@ -4,49 +4,25 @@
-const GLOBAL_TRACER_SYMBOL = Symbol.for("@azure/core-tracing.tracerCache3");
+// tracerCache will be updated when a new incompatible version of OTel is
+// shipped in core-tracing.
+const GLOBAL_TRACER_SYMBOL = Symbol.for(`@azure/core-tracing.tracerCache4`);
```
I think if we keep the simple numbers we should leave comments about what numbers correspond with which minimum OT package requirements, otherwise this will be annoying to dig out of git history someday.

maorleger replied:
Done! I added tracerCache3 and tracerCache4 comments. For tracerCache 1 and 2 history is little muddier so wasn't easy to track down.

xirzec commented on sdk/core/core-tracing/test/cache.spec.ts:
```diff
@@ -0,0 +1,25 @@
+import * as fs from "fs";
+import * as path from "path";
+
+const validOpenTelemetryVersions = ["1.0.0-rc1", "0.20.0"];
```
I just noticed that it looks like 1.0 has shipped: https://www.npmjs.com/package/@opentelemetry/api/v/1.0.0

Should we just update to that and GA this package instead? :)

ramya-rao-a replied:
Oh wow! Oh wow! Oh wow!

ramya-rao-a replied:
But 1.0.0 has the next tag... and 0.20.0 has the latest tag

xirzec reviewed (commented):
Left a code change comment, but also just made an interesting discovery...

maorleger commented:
@xirzec I addressed the feedback and implemented the changes we discussed offline (removal of the version collision runtime check and adding a sanity check in test to ensure future us remembers to think about compatibility) - any concerns with the latest changes? If not, should we merge this in?

Especially wanted to call out the `cache.ts` changes which, while simplifying the code, could use some extra scrutiny 😄

xirzec commented on sdk/core/core-http/test/policies/tracingPolicyTests.ts:
```diff
@@ -12,32 +12,49 @@ import {
 import { tracingPolicy } from "../../src/policies/tracingPolicy";
 
-class MockSpan extends NoOpSpan {
+class MockSpan implements Span {
```
much simpler! Good discovery here

xirzec reviewed (approved):
Bravo! 👏
